Add timeout fine-tunning advanced docs section (#476)
encode/httpx: docs/advanced.md

# This request will be tunneled instead of forwarded.
client.get("http://example.com")
```

## Timeout fine-tuning

> HTTPX library offers many request timeout management options. HTTPX distinguish three types of timeouts: **connect** timeouts, 

Reviewer: Phrasing suggestion:

```suggestion
HTTPX offers various request timeout management options. Three types of timeouts are available: **connect** timeouts, 
```

**write** timeouts and **read** timeouts.

> * Connect timeout specify time period within which a connection between a client and a server must be established.

Reviewer: Phrasing suggestion, and also an idea about documenting what happens in case of timeout:

```suggestion
* The **connect timeout** specifies the maximum amount of time to wait until a connection to the requested host is established. If HTTPX is unable to connect within this time frame, a `ConnectTimeout` exception is raised.
```

> * Write timeout specify the maximum duration before timing out writes of the response.

Reviewer: Actually, the write timeout applies between the sending of individual *chunks* of *request* data:

```suggestion
* Write timeout specifies the maximum duration to wait for a chunk of data to be sent (for example, a chunk of the request body). If HTTPX is unable to send data within this time frame, a `WriteTimeout` exception is raised.
```

Hope I'm not mistaken that *writing* refer to *sending a request*. Makes me think that connect/sending/receive might be more intuitive wording perhaps?

> * Read timeout specify the maximum duration for reading the entire request.

Reviewer: And here, it turns out that the read timeout applies to receiving individual *chunks* of the *response*:

```suggestion
* Read timeout specifies the maximum duration to wait for a chunk of data to be received (for example, a chunk of the response body). If HTTPX is unable to receive data within this time frame, a `ReadTimeout` exception is raised.
```

HTTPX allows you to manage timeout policy freely. You can set timeout parameter at client creation phase or pass it on every request.  

> In a nutshell, user can manage timeouts on three different levels of abstraction:

Reviewer: It seems this part confuses how to define the *scope* of the timeout (client-wide vs. per-request) and how to define its *value* (e.g. as an integer vs. a `TimeoutConfig` instance)

I think we can split things up like this:

```markdown
### Setting timeouts

You can set timeouts on two levels:

- For a given request:

<!-- "Request level" example -->

- On a client instance, which results in the given `timeout` being used as a default for requests made with this client:

<!-- "Client level" example -->

Besides, you can pass timeouts in two forms:

- A number, which sets the read, write and connect timeouts to the same value, as in the examples above.
- A `TimeoutConfig` instance, which allows to define the read, write and connect timeouts independently:

<!-- Example -->
```

**Request level:**
```python

> httpx.get('http://example.com/api/v1/example', timeout=5)

Reviewer: Probably worth adding an example with a `Client` here as well:

```suggestion
httpx.get('http://example.com/api/v1/example', timeout=5)
# Or, with a client:
client = httpx.Client()
client.get("http://example.com/api/v1/example", timeout=5)
```

```

**Client level:**
```python
client = httpx.Client(timeout=5)
client.get('http://example.com/api/v1/example')
```

**TimeoutConfig level:**  
Unlike request and client level configuration, TimeoutConfig object gives the ability to configure
every distinct timeout type separately.
```python
timeout = httpx.TimeoutConfig(
    connect_timeout=5, 
    read_timeout=5, 
    write_timeout=5
)

resp = httpx.get('http://example.com/api/v1/example', timeout=timeout)
```

> By default all types of timeouts are set to 5 second, in order to **disable** them you have to pass ``None`` as timeout parameter.

Reviewer: Should we split this information into sub-sections as well for clarity?

```markdown
### Default timeouts

By default, all types of timeouts are set to **5 seconds**.

### Disabling timeouts

To disable timeouts, you can pass `None` as a timeout parameter. Note that currently this is not supported by the top-level API.

...
```

Author: I also was thinking about it. Fixed
